File: Python/day24/AOC.py
import re
from itertools import combinations
from collections import namedtuple

Gate = namedtuple("Gate", ["op", "left", "right", "output"])

OPERATORS = {
    "OR": lambda x, y: x | y,
    "AND": lambda x, y: x & y,
    "XOR": lambda x, y: x ^ y,
}

def parse_input(input_data):
    parts = input_data.split("\n\n")
    pattern = r"(.+?)\s+(AND|OR|XOR)\s+(.+?)\s+->\s+(.+)"
    values, gates = {}, []
    for part in parts[0].split("\n"):
        key, value = part.split(":")
        values[key] = int(value)   
    for part in parts[1].split("\n"):
        left, op, right, output = re.match(pattern, part).groups()
        gates.append(Gate(op, left, right, output))
    return values, gates


def simulate_system(values, gates):
    wires = dict(values)
    unresolved_gates = gates[:]

    while unresolved_gates:
        for g in gates:
            if g.output in wires:
                continue
            if g.left in wires and g.right in wires:
                wires[g.output] = OPERATORS[g.op](wires[g.left], wires[g.right])
                unresolved_gates.remove(g)
    return wires

def calculate_output(wires):
    z_wires = sorted(((k, v) for k, v in wires.items() if k.startswith("z")), reverse=True)
    return int("".join(str(v[1]) for v in z_wires), 2)

def part1(input):
    return calculate_output(simulate_system(*parse_input(input))) 

# We are dealing with a ripple-carry adder here (https://www.circuitstoday.com/ripple-carry-adder).
# It adds two 45-bit numbers, x and y, and outputs a 46-bit number z. The first 45 bits of z are the sum of the corresponding bits in x and y, and the last bit is the carry out.
# Illustrations of what can be found in this folder and here: https://www.reddit.com/r/adventofcode/comments/1hl8tl4/2024_day_24_part_2_using_a_graph_visualization/
# Bascially, we "traverse" the graph of operations and check if we can get to the end. If we can't, we try to swap two outputs and see if we can get further.
def check_ripple(gates):
    ops = {}
    for op, x1, x2, output in gates:
        ops[(frozenset([x1, x2]), op)] = output
    def get_output(x1, x2, op):
        return ops.get((frozenset([x1, x2]), op), None)
    carries = {}
    correct = set()
    prev_intermediates = set()
    for i in range(45):
        pos = f"0{i}" if i < 10 else str(i)
        predigit = get_output(f"x{pos}", f"y{pos}", "XOR")
        precarry1 = get_output(f"x{pos}", f"y{pos}", "AND")
        if i == 0:
            carries[i] = precarry1
            continue
        digit = get_output(carries[i - 1], predigit, "XOR")
        if digit != f"z{pos}":
            return i - 1, correct
        correct.add(carries[i - 1])
        correct.add(predigit)
        for wire in prev_intermediates:
            correct.add(wire)
        precarry2 = get_output(carries[i - 1], predigit, "AND")
        carry_out = get_output(precarry1, precarry2, "OR")
        carries[i] = carry_out
        prev_intermediates = set([precarry1, precarry2])
    return i, correct

def part2(input):
    _, gates = parse_input(input)
    swaps = []

    base, base_used = check_ripple(gates)
    for _ in range(4):
        for i, j in combinations(range(len(gates)), 2):
            op_i, left_i, right_i, ouput_i = gates[i]
            op_j, left_j, right_j, output_j = gates[j]
            if "z00" in (ouput_i, output_j):
                continue
            if ouput_i in base_used or output_j in base_used:
                continue
            gates[i] = op_i, left_i, right_i, output_j
            gates[j] = op_j, left_j, right_j, ouput_i
            attempt, attempt_used = check_ripple(gates)
            if attempt > base:
                swaps.extend([ouput_i, output_j])
                base, base_used = attempt, attempt_used
                break
            gates[i] = op_i, left_i, right_i, ouput_i
            gates[j] = op_j, left_j, right_j, output_j

    return ",".join(sorted(swaps))

def main():
    input = open("input.txt", "r").read()
    print(f"Part 1: {part1(input)}")
    print(f"Part 2: {part2(input)}")

if __name__ == "__main__":
    main()

File: Go/day24/AOC.go
package main

import (
	"fmt"
	"os"
	"sort"
	"strconv"
	"strings"
	"utils"
)

type Gate struct{ op, left, right, output string }

var OPERATIONS = map[string]func(int, int) int{
	"OR":  func(a, b int) int { return a | b },
	"AND": func(a, b int) int { return a & b },
	"XOR": func(a, b int) int { return a ^ b },
}

func parseInput(input string) (map[string]int, []Gate) {
	wires := map[string]int{}
	gates := []Gate{}

	parts := strings.Split(input, "\n\n")
	for _, line := range strings.Split(parts[0], "\n") {
		p := strings.Split(line, ": ")
		wires[p[0]] = utils.Int(p[1])
	}

	for _, line := range strings.Split(parts[1], "\n") {
		p := strings.Split(line, " ")
		gates = append(gates, Gate{p[1], p[0], p[2], p[4]})
	}

	return wires, gates
}

func eval(wires map[string]int, gates []Gate) map[string]int {
	for {
		changed := false
		for _, gate := range gates {
			if _, ok := wires[gate.output]; ok {
				continue
			}

			left, left_ok := wires[gate.left]
			right, right_ok := wires[gate.right]

			if left_ok && right_ok {
				wires[gate.output] = OPERATIONS[gate.op](left, right)
				changed = true
			}
		}

		if !changed {
			break
		}
	}

	return wires
}

func part1(input string) int {
	wires, gates := parseInput(input)
	wires = eval(wires, gates)

	zWires := []int{}
	for i := 0; i < len(wires); i++ {
		if z, ok := wires[fmt.Sprintf("z%02d", i)]; ok {
			zWires = append(zWires, z)
		} else {
			break
		}
	}
	// reverse the wires
	for i, j := 0, len(zWires)-1; i < j; i, j = i+1, j-1 {
		zWires[i], zWires[j] = zWires[j], zWires[i]
	}
	// convert to decimal
	str := utils.Apply(zWires, func(x int) string { return strconv.Itoa(x) })
	decimal, _ := strconv.ParseInt(strings.Join(str, ""), 2, 64)
	return int(decimal)
}

func checkRippleCarry(gates []Gate) (int, utils.Set[string]) {
	ops := make(map[string]string)
	for _, gate := range gates {
		ops[gate.left+"|"+gate.right+"|"+gate.op] = gate.output
		ops[gate.right+"|"+gate.left+"|"+gate.op] = gate.output
	}
	carries := make([]string, 45)
	correct := utils.NewSet[string]()
	prevIntermediates := utils.NewSet[string]()

	for i := 0; i < 45; i++ {
		pos := fmt.Sprintf("%02d", i)
		predigit := ops["x"+pos+"|y"+pos+"|XOR"]
		precarry1 := ops["x"+pos+"|y"+pos+"|AND"]
		if i == 0 {
			carries[i] = precarry1
			continue
		}
		digit := ops[carries[i-1]+"|"+predigit+"|XOR"]
		if digit != fmt.Sprintf("z%s", pos) {
			return i - 1, correct
		}
		correct.Add(carries[i-1])
		correct.Add(predigit)
		for wire := range prevIntermediates {
			correct.Add(wire)
		}
		precarry2 := ops[carries[i-1]+"|"+predigit+"|AND"]
		carryOut := ops[precarry1+"|"+precarry2+"|OR"]
		carries[i] = carryOut
		prevIntermediates.Add(precarry1)
		prevIntermediates.Add(precarry2)
	}
	return 45, correct
}

func pairs(n int) [][2]int {
	result := make([][2]int, 0, n*(n-1)/2)
	for i := 0; i < n; i++ {
		for j := i + 1; j < n; j++ {
			result = append(result, [2]int{i, j})
		}
	}
	return result
}

func part2(input string) string {
	_, gates := parseInput(input)
	swaps := []string{}

	base, baseUsed := checkRippleCarry(gates)
	for i := 0; i < 4; i++ {
		for _, p := range pairs(len(gates)) {
			x, y := p[0], p[1]

			if gates[x].output == "z00" || gates[y].output == "z00" {
				continue
			}

			if _, ok := baseUsed[gates[x].output]; ok {
				continue
			}

			if _, ok := baseUsed[gates[y].output]; ok {
				continue
			}

			gates[x].output, gates[y].output = gates[y].output, gates[x].output
			attempt, used := checkRippleCarry(gates)
			if attempt > base {
				swaps = append(swaps, gates[x].output, gates[y].output)
				base = attempt
				baseUsed = used
				break
			}
			gates[x].output, gates[y].output = gates[y].output, gates[x].output
		}
	}

	sort.Strings(swaps)
	return strings.Join(swaps, ",")
}

func main() {
	input, err := os.ReadFile("input.txt")
	if err != nil {
		fmt.Print("Error reading file\n")
		panic(err)
	}

	lines := string(input)
	part1_ans := part1(lines)
	fmt.Printf("Part 1: %d\n", part1_ans)

	part2_ans := part2(lines)
	fmt.Printf("Part 2: %s\n", part2_ans)
}
